deepspeed/moe/sharded_moe.py

```diff
@@ -322,7 +322,7 @@ def top2gating(logits: Tensor,
     l_aux = torch.mean(me * ce) * num_experts * num_experts
 
     # gating decisions
-    exp_counts = torch.sum(mask1 + mask2, dim=0)
+    exp_counts = torch.sum(mask1 + mask2, dim=0).detach().to(logits.device)
```

Comment:
Can the device of `mask1` and `mask1` be different from `logits`?

Reply:
> Can the device of `mask1` and `mask1` be different from `logits`?

From[ line 296 to 301](https://github.com/microsoft/DeepSpeed/pull/5578/files#diff-7d28ff32544f36339d99fdf142c2ab63d7332305758be0ff7ccc94e6350e1476R296-R301), we can find that , the calculation of `mask1` depends on `logits`, and all torch operation will keep the original device,  so the device of `mask1` and `logits` be the same one. The same to `mask1` and `mask2` [line 309 to 311](https://github.com/microsoft/DeepSpeed/pull/5578/files#diff-7d28ff32544f36339d99fdf142c2ab63d7332305758be0ff7ccc94e6350e1476R309-R311)

Reply:
This causes no harm, and the loss remains consistent. By removing this CPU sync, it will improve performance by 1-2% on the GPU and 5%+ on the HPU. Could you please take a look again ?  @tohtana Thanks

Reply:
Thank you for the clarification!